PXC-3464: Data is not propagated with SET SESSION sql_log_bin = 0

kamil-holubicki opened the pull request:
https://jira.percona.com/browse/PXC-3464

Problem:
When sql_log_bin session variable is set to OFF, queries are not
replicated across PXC cluster nodes

Cause:
PXC flow decides if particular statement needs to be replicated basing
on inter alia sql_log_bin value.

Solution:
All statements are replicated. However we are still blocking replication
if it was blocked internally by the server (auxiliary
OPTION_BIN_LOG_INTERNAL_OFF flag introduced)

dutow commented on sql/wsrep_schema.cc:
```diff
@@ -108,12 +108,14 @@ class binlog_off {
       : m_thd(thd),
         m_option_bits(thd->variables.option_bits),
         m_sql_log_bin(thd->variables.sql_log_bin) {
+    thd->set_sql_log_bin_off();
```
this could use a `log_bin_off_guard` data member instead?

dutow commented on sql/thd_raii.h:
```diff
@@ -179,6 +179,7 @@ class Disable_binlog_guard {
         m_wsrep_on(thd->variables.wsrep_on),
         m_binlog_disabled(thd->variables.option_bits & OPTION_BIN_LOG) {
     thd->variables.option_bits &= ~OPTION_BIN_LOG;
+    thd->set_sql_log_bin_off();
```
This could also use a `log_bin_off_guard` data member

dutow commented on sql/sql_class.cc:
```diff
@@ -237,6 +237,7 @@ THD::Attachable_trx::Attachable_trx(THD *thd, Attachable_trx *prev_trx)
   // Nothing should be binlogged from attachable transactions and disabling
   // the binary log allows skipping some code related to figuring out what
   // log format should be used.
+  m_thd->set_sql_log_bin_off();
```
This could also use a `log_bin_off_guard` data member instead.

dutow commented on sql/rpl_info_table.cc:
```diff
@@ -118,6 +118,7 @@ int Rpl_info_table::do_init_info(enum_find_method method, uint instance) {
   THD *thd = access->create_thd();
 
   saved_mode = thd->variables.sql_mode;
+  thd->set_sql_log_bin_off();
```
Why not a `log_bin_off_guard`?

kamil-holubicki replied:
because of thd lifetime. If we used RAII here, guard is destroyed after the thd.

dutow commented on sql/dd/impl/transaction_impl.cc:
```diff
@@ -162,6 +162,7 @@ Update_dictionary_tables_ctx::Update_dictionary_tables_ctx(THD *thd)
     m_thd->clear_current_stmt_binlog_format_row();
 
   // Disable bin logging
+  m_thd->set_sql_log_bin_off();
```
This could use a `log_bin_off_guard` data member.

kamil-holubicki replied:
Will be reworked

dutow commented on sql/rpl_info_table.cc:
```diff
@@ -184,6 +186,7 @@ int Rpl_info_table::do_flush_info(const bool force) {
 
   sync_counter = 0;
   saved_mode = thd->variables.sql_mode;
+  thd->set_sql_log_bin_off();
```
Why not a `log_bin_off_guard`?

kamil-holubicki replied:
because of thd lifetime. If we used RAII here, guard is destroyed after the thd.

dutow commented on sql/sql_class.h:
```diff
@@ -4955,6 +4955,40 @@ class THD : public MDL_context_owner,
+  }
+
+  bool sql_log_bin_off() const {
+      DBUG_ASSERT(sql_log_bin_off_saved >= 0);
```
I would also add an assert to the destructor ensuring that it is 0.

dutow commented on sql/sql_class.h:
```diff
@@ -4955,6 +4955,40 @@ class THD : public MDL_context_owner,
+
+class log_bin_off_guard {
+  public:
+    log_bin_off_guard(THD* thd)
```
explicit

kennt-percona commented on mysql-test/suite/galera/t/session_logbin_off.test:
```diff
@@ -0,0 +1,76 @@
+DELETE FROM t1;
+
+--connection node_2
+--let $assert_text = node_2 t1 should have zero rows
```
You may want to make this into a wait condition.  This assert may fail under load if the DELETE is not replicated in time.  Similar situation below also.

kamil-holubicki replied:
Good point. I will rework in the same way for 5.7.

kamil-holubicki committed:
PXC-3464: Data is not propagated with
SET SESSION sql_log_bin = 0

https://jira.percona.com/browse/PXC-3464

Problem:
When sql_log_bin session variable is set to OFF, queries are not
replicated across PXC cluster nodes

Cause:
PXC flow decides if particular statement needs to be replicated basing
on inter alia sql_log_bin value.

Solution:
All statements are replicated. However we are still blocking replication
if it was blocked internally by the server (auxiliary
OPTION_BIN_LOG_INTERNAL_OFF flag introduced)

kamil-holubicki commented:
I have reworked it completely and now it follows the same way it is done on 5.7.

kamil-holubicki commented:
https://pxc.cd.percona.com/view/PXC%208.0/job/pxc-8.0-param/212
No failures caused by this fix.

venkatesh-prasad-v reviewed (approved):
Thank you for working on this. Approved!
